Add support for the alarms Web Extension API.

xeenon opened the pull request:
#### de4ab2c5684f9ed12ea189d1d30aa1499c1207e6
<pre>
Add support for the alarms Web Extension API.
<a href="https://webkit.org/b/259950">https://webkit.org/b/259950</a>
rdar://problem/113592049

Reviewed by Brian Weinstein.

Implements the browser.alarms namespace and added API tests for all the supported methods.
Documentation: <a href="https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/alarms">https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/alarms</a>

* Source/WebKit/DerivedSources-input.xcfilelist: Added alarm files.
* Source/WebKit/DerivedSources-output.xcfilelist: Ditto.
* Source/WebKit/DerivedSources.make: Added WebExtensionAPIAlarms interface.
* Source/WebKit/Shared/Extensions/WebExtensionAlarmParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionAlarmParameters.serialization.in: Added.
* Source/WebKit/Sources.txt: Added alarm files.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm: Added.
(WebKit::WebExtensionContext::alarmsCreate):
(WebKit::WebExtensionContext::alarmsGet):
(WebKit::WebExtensionContext::alarmsClear):
(WebKit::WebExtensionContext::alarmsGetAll):
(WebKit::WebExtensionContext::alarmsClearAll):
(WebKit::WebExtensionContext::fireAlarmsEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtensionAlarm.cpp: Added.
(WebKit::WebExtensionAlarm::schedule):
(WebKit::WebExtensionAlarm::fire):
* Source/WebKit/UIProcess/Extensions/WebExtensionAlarm.h: Added.
(WebKit::WebExtensionAlarm::create):
(WebKit::WebExtensionAlarm::WebExtensionAlarm):
(WebKit::WebExtensionAlarm::parameters const):
(WebKit::WebExtensionAlarm::name const):
(WebKit::WebExtensionAlarm::initialInterval const):
(WebKit::WebExtensionAlarm::repeatInterval const):
(WebKit::WebExtensionAlarm::nextScheduledTime const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in: Added alarm messages.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm: Added.
(WebKit::toAPI):
(WebKit::WebExtensionAPIAlarms::createAlarm):
(WebKit::WebExtensionAPIAlarms::get):
(WebKit::WebExtensionAPIAlarms::getAll):
(WebKit::WebExtensionAPIAlarms::clear):
(WebKit::WebExtensionAPIAlarms::clearAll):
(WebKit::WebExtensionAPIAlarms::onAlarm):
(WebKit::WebExtensionContextProxy::dispatchAlarmEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Added FIXME and check permissions for alarms too.
(WebKit::WebExtensionAPINamespace::alarms): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIAlarms.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIAlarms.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Added alarms.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in: Added alarm messages.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAlarms.mm: Added.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
(TestWebKitAPI::Util::constructScript): Stop adding semicolons, it can cause mysterious errors.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm: Use &quot;UIKitSPI.h&quot; since it was conflicting
with &lt;pal/spi/ios/UIKitSPI.h&gt; due to unified sources shifting around.
* Tools/TestWebKitAPI/ios/UIKitSPI.h: Add missing interfaces needed by WKWebViewGetContents.mm.

Canonical link: <a href="https://commits.webkit.org/266742@main">https://commits.webkit.org/266742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b23b72598d032612112fa4410f61b8cf5c0460f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16388 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17467 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15033 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14821 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/15330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/12438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/17122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/12613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/13205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/17122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/13692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/13371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/17122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/13924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/13218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/17556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1751 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/13767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
```diff
@@ -0,0 +1,91 @@
+void WebExtensionContext::alarmsGetAll(CompletionHandler<void(Vector<WebExtensionAlarmParameters>&&)>&& completionHandler)
+{
+    Vector<WebExtensionAlarmParameters> alarms;
+    alarms.reserveCapacity(m_alarmMap.size());
```
is `reserveCapacity` preferred over `reserveInitialCapacity`, which we've used in the past? Also not sure what the difference is between the two

xeenon replied:
Ah this was meant to be `reserveInitialCapacity`. Thanks!

b-weinstein commented on Source/WebKit/UIProcess/Extensions/WebExtensionAlarm.cpp:
```diff
@@ -0,0 +1,62 @@
+
+void WebExtensionAlarm::fire()
+{
+    // Calculate the next scheduled time now, so the handler's work time does not count against.
```
Nit: `does not count against it`?

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm:
```diff
@@ -0,0 +1,203 @@
+
+static inline NSDictionary *toAPI(const WebExtensionAlarmParameters& alarm)
+{
+    NSMutableDictionary *result = [NSMutableDictionary dictionaryWithCapacity:3];
```
Super-nit: Should this be 2 since the only guaranteed entries are name and scheduledTime?

xeenon replied:
Yeah I was on the fence about this. It is almost always going to be 3, since alarms usually repeat.

b-weinstein commented on Tools/TestWebKitAPI/ios/UIKitSPI.h:
```diff
@@ -351,6 +351,25 @@ typedef NS_ENUM(NSUInteger, _UIClickInteractionEvent) {
+@interface NSParagraphStyle ()
+- (NSArray<NSTextBlock *> *)textBlocks;
+@end
+
```
What made this required?

xeenon replied:
Unified source made the build fail when I added a new file. Mentioned this in the ChangeLog, near the end.

xeenon committed:
Add support for the alarms Web Extension API.
https://webkit.org/b/259950
rdar://problem/113592049

Reviewed by Brian Weinstein.

Implements the browser.alarms namespace and added API tests for all the supported methods.
Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/alarms

* Source/WebKit/DerivedSources-input.xcfilelist: Added alarm files.
* Source/WebKit/DerivedSources-output.xcfilelist: Ditto.
* Source/WebKit/DerivedSources.make: Added WebExtensionAPIAlarms interface.
* Source/WebKit/Shared/Extensions/WebExtensionAlarmParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionAlarmParameters.serialization.in: Added.
* Source/WebKit/Sources.txt: Added alarm files.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm: Added.
(WebKit::WebExtensionContext::alarmsCreate):
(WebKit::WebExtensionContext::alarmsGet):
(WebKit::WebExtensionContext::alarmsClear):
(WebKit::WebExtensionContext::alarmsGetAll):
(WebKit::WebExtensionContext::alarmsClearAll):
(WebKit::WebExtensionContext::fireAlarmsEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtensionAlarm.cpp: Added.
(WebKit::WebExtensionAlarm::schedule):
(WebKit::WebExtensionAlarm::fire):
* Source/WebKit/UIProcess/Extensions/WebExtensionAlarm.h: Added.
(WebKit::WebExtensionAlarm::create):
(WebKit::WebExtensionAlarm::WebExtensionAlarm):
(WebKit::WebExtensionAlarm::parameters const):
(WebKit::WebExtensionAlarm::name const):
(WebKit::WebExtensionAlarm::initialInterval const):
(WebKit::WebExtensionAlarm::repeatInterval const):
(WebKit::WebExtensionAlarm::nextScheduledTime const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in: Added alarm messages.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm: Added.
(WebKit::toAPI):
(WebKit::WebExtensionAPIAlarms::createAlarm):
(WebKit::WebExtensionAPIAlarms::get):
(WebKit::WebExtensionAPIAlarms::getAll):
(WebKit::WebExtensionAPIAlarms::clear):
(WebKit::WebExtensionAPIAlarms::clearAll):
(WebKit::WebExtensionAPIAlarms::onAlarm):
(WebKit::WebExtensionContextProxy::dispatchAlarmEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Added FIXME and check permissions for alarms too.
(WebKit::WebExtensionAPINamespace::alarms): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIAlarms.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIAlarms.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Added alarms.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in: Added alarm messages.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAlarms.mm: Added.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
(TestWebKitAPI::Util::constructScript): Stop adding semicolons, it can cause mysterious errors.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm: Use "UIKitSPI.h" since it was conflicting
with <pal/spi/ios/UIKitSPI.h> due to unified sources shifting around.
* Tools/TestWebKitAPI/ios/UIKitSPI.h: Add missing interfaces needed by WKWebViewGetContents.mm.

Canonical link: https://commits.webkit.org/266742@main
